### PR TITLE
Add "gem rebuild" command.

### PR DESCRIPTION
**Pull request by duckinator**
## What was the end-user or developer problem that led to this PR?

If a gem implements the requirements for reproducible builds, and I have a copy of the source, I want to be able to verify that the released version was built from that source.

## What is your fix for the problem, implemented in this PR?

I implemented the `gem rebuild` command. See the discussion in #3118 for details.

```
$ ls -a
./  ../
$ git clone https://github.com/duckinator/okay.git
Cloning into 'okay'...
remote: Enumerating objects: 483, done.
remote: Counting objects: 100% (169/169), done.
remote: Compressing objects: 100% (86/86), done.
remote: Total 483 (delta 87), reused 134 (delta 83), pack-reused 314
Receiving objects: 100% (483/483), 77.50 KiB | 3.52 MiB/s, done.
Resolving deltas: 100% (213/213), done.
$ gem rebuild -C ./okay okay 12.0.1
Fetching okay-12.0.1.gem
Downloaded okay version 12.0.1 as /usr/home/puppy/dev/rebuild-test/rebuild/old/okay-12.0.1.gem.
WARNING:  open-ended dependency on cacert (>= 0) is not recommended
  use a bounded requirement, such as '~> x.y'
WARNING:  See https://guides.rubygems.org/specification-reference/ for help
  Successfully built RubyGem
  Name: okay
  Version: 12.0.1
  File: okay-12.0.1.gem

Built at: 2021-09-11 20:01:16 EDT (1631404876)
Original build saved to:   /usr/home/puppy/dev/rebuild-test/rebuild/old/okay-12.0.1.gem
Reproduced build saved to: /usr/home/puppy/dev/rebuild-test/rebuild/new/okay-12.0.1.gem
Working directory: ./okay

Hash comparison:
  558bf6a6343c984f11021aff41c1a63ec07911c0658a408af6e4308b3e7764ea      /usr/home/puppy/dev/rebuild-test/rebuild/old/okay-12.0.1.gem
  558bf6a6343c984f11021aff41c1a63ec07911c0658a408af6e4308b3e7764ea      /usr/home/puppy/dev/rebuild-test/rebuild/new/okay-12.0.1.gem

SUCCESS - original and rebuild hashes matched
$
```

Failed example:

```
$ ls -a
./  ../
$ git clone https://github.com/duckinator/okay.git
Cloning into 'okay'...
remote: Enumerating objects: 483, done.
remote: Counting objects: 100% (169/169), done.
remote: Compressing objects: 100% (86/86), done.
remote: Total 483 (delta 87), reused 134 (delta 83), pack-reused 314
Receiving objects: 100% (483/483), 77.50 KiB | 3.10 MiB/s, done.
Resolving deltas: 100% (213/213), done.
$ echo '#hi' >> okay/lib/okay.rb 
$ tail -n 2 okay/lib/okay.rb 
end
#hi
$ gem rebuild -C ./okay okay 12.0.1
Fetching okay-12.0.1.gem
Downloaded okay version 12.0.1 as /usr/home/puppy/dev/rebuild-test/rebuild/old/okay-12.0.1.gem.
WARNING:  open-ended dependency on cacert (>= 0) is not recommended
  use a bounded requirement, such as '~> x.y'
WARNING:  See https://guides.rubygems.org/specification-reference/ for help
  Successfully built RubyGem
  Name: okay
  Version: 12.0.1
  File: okay-12.0.1.gem

Built at: 2021-09-11 20:01:16 EDT (1631404876)
Original build saved to:   /usr/home/puppy/dev/rebuild-test/rebuild/old/okay-12.0.1.gem
Reproduced build saved to: /usr/home/puppy/dev/rebuild-test/rebuild/new/okay-12.0.1.gem
Working directory: ./okay

Hash comparison:
  558bf6a6343c984f11021aff41c1a63ec07911c0658a408af6e4308b3e7764ea      /usr/home/puppy/dev/rebuild-test/rebuild/old/okay-12.0.1.gem
  f803f93864366942b9aed3e5f7d806c1bffccae5d59eff74f68dbd59d6811426      /usr/home/puppy/dev/rebuild-test/rebuild/new/okay-12.0.1.gem

FAILURE - original and rebuild hashes did not match
$
```

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by duckinator** [comment]
TODO:

- [ ] add tests
- [ ] update the example in the help text (i changed the output format since i wrote it)

**Comment by matiaskorhonen** [comment]
Does this work with cryptographically signed gems? (see https://guides.rubygems.org/security/)

**Comment by duckinator** [comment]
@matiaskorhonen I haven't put in work to specifically support that, so I'd be surprised if it worked. I'm not sure what would be needed to support it.

**Comment by duckinator** [comment]
Sorry for the slow movement on this &mdash; I've been struggling with basically-perpetual brain fog the last two months or so, and it's making it really difficult to actually work on the RubyGems codebase.

I did finally get this to the point that it has a test which fails _when expected to_ instead of passing unexpectedly, which is nice.

At the moment, the test fails because it tries to download something not on RubyGems.org. I tried using the `--original` flag to specify a local file to compare against but it just gave a different error instead (see the TODO near the end of `test/rubygems/test_gem_commands_rebuild_command.rb`). That's the next thing to figure out, I think.

**Comment by duckinator** [comment]
Rebased off master.

**Comment by duckinator** [comment]
Rebased off master (d4325048dce3ae4d5d63a3efa7a776a1c2d4d2eb).

I still need to resolve this: https://github.com/rubygems/rubygems/blob/fbe6fcc75ed61278ccca55fabbfd2edbf4d1eb29/test/rubygems/test_gem_commands_rebuild_command.rb#L114-L115

**Comment by duckinator** [comment]
Rebased off master (c44fe8e).

**Comment by duckinator** [comment]
Resolved linter errors.

**Comment by duckinator** [comment]
I'm having trouble figuring out why the tests are failing, but the `gem rebuild` command itself is working:

```
~/test/rebuild$ ls -a
./  ../
~/test/rebuild$ git clone --branch v12.0.2 https://github.com/duckinator/okay.git
Cloning into 'okay'...
remote: Enumerating objects: 512, done.
remote: Counting objects: 100% (43/43), done.
remote: Compressing objects: 100% (43/43), done.
remote: Total 512 (delta 0), reused 0 (delta 0), pack-reused 469
Receiving objects: 100% (512/512), 73.46 KiB | 2.62 MiB/s, done.
Resolving deltas: 100% (240/240), done.
~/test/rebuild$ gem rebuild -C ./okay --gemspec okay.gemspec okay 12.0.2
Fetching okay-12.0.2.gem
Downloaded okay version 12.0.2 as /usr/home/puppy/test/rebuild/rebuild/old/okay-12.0.2.gem.
WARNING:  open-ended dependency on cacert (>= 0) is not recommended
  use a bounded requirement, such as "~> x.y"
WARNING:  See https://guides.rubygems.org/specification-reference/ for help
  Successfully built RubyGem
  Name: okay
  Version: 12.0.2
  File: okay-12.0.2.gem

Built at: 2023-08-31 21:39:02 EDT (1693532342)
Original build saved to:   /usr/home/puppy/test/rebuild/rebuild/old/okay-12.0.2.gem
Reproduced build saved to: /usr/home/puppy/test/rebuild/rebuild/new/okay-12.0.2.gem
Working directory: ./okay

Hash comparison:
  38a8bd78ce10bc19189ead0b56fa490d03788a2f926a6481b2f1f5d5fa5ab75b      /usr/home/puppy/test/rebuild/rebuild/old/okay-12.0.2.gem
  38a8bd78ce10bc19189ead0b56fa490d03788a2f926a6481b2f1f5d5fa5ab75b      /usr/home/puppy/test/rebuild/rebuild/new/okay-12.0.2.gem

SUCCESS - original and rebuild hashes matched
~/test/rebuild$
```

**Comment by duckinator** [comment]
Problems that have been resolved:

- the test does not do `sleep 1` to ensure time differences anymore
- there is no longer two almost-identical complex functions
- the test gives the `gem rebase` command the original gem file, now

So at this point I think the _only_ problem is that the test doesn't actually run.

The fact that it's eating stderr is making this a lot more difficult to figure out.

**Comment by deivid-rodriguez** [comment]
Thanks for the hard work @duckinator. Let me know if you want me to have a look at the failing tests!

**Comment by duckinator** [comment]
@deivid-rodriguez if you could take a look, I'd definitely appreciate it. They both fail at the same spot (the `rebuild_gem_file = util_test_rebuild_gem(...)` line), but I have no idea why.

**Comment by duckinator** [comment]
I _finally_ figured out what the problem is. (Or at least part of it, anyway.)

My code does `find_gemspec("#{gem_name}.spec")` when no gemspec is provided.

It turns out that `util_spec` creates `#{gem_name}-#{gem_version}.spec`.

So I can just specify the exact path and it should get past that.

(I don't want to even think about how much time I've spent debugging this...)

**Comment by duckinator** [comment]
I am extremely excited because after TWO AND A HALF FREAKING YEARS, if CI passes, then this is FINALLY ready for a proper review.

**Comment by deivid-rodriguez** [comment]
@duckinator So sorry I forgot to look into this, I'm glad that you managed to fix the remaining issues! 💪

File: lib/rubygems/commands/rebuild_command.rb
Comment (segiddins) [inline lib/rubygems/commands/rebuild_command.rb:150]: should we add an option to maybe run [`diffoscope`](https://diffoscope.org) if present against the two files, to try and help reveal the differences between the two files?

File: lib/rubygems/commands/rebuild_command.rb
Comment (segiddins) [inline lib/rubygems/commands/rebuild_command.rb:121]: note to self: this is a sign we should embed the source date epoch used into the metadata of the built gemspec. additionally, ruby/rubygems versions would be helpful

File: lib/rubygems/commands/rebuild_command.rb
Comment (segiddins) [inline lib/rubygems/commands/rebuild_command.rb:123]: I think this will leak the files, use the block form of `File.open` with `"rb"` ?

File: lib/rubygems/commands/rebuild_command.rb
Comment (segiddins) [inline lib/rubygems/commands/rebuild_command.rb:160]: this will cause errors if `rebuild` is a file and not a directory I think

File: lib/rubygems/commands/rebuild_command.rb
Comment (segiddins) [inline lib/rubygems/commands/rebuild_command.rb:220]: can this share an implementation with the `build` command?

File: lib/rubygems/commands/rebuild_command.rb
Comment (deivid-rodriguez) [reply lib/rubygems/commands/rebuild_command.rb:150]: This seems like a good idea!

File: lib/rubygems/commands/rebuild_command.rb
Comment (deivid-rodriguez) [inline lib/rubygems/commands/rebuild_command.rb:55]: I think, since the running RubyGems version is [included in metadata](https://github.com/rubygems/rubygems/blob/a7aa64a918a590efe0e8c4da0eba189a8aa5140d/lib/rubygems/specification.rb#L153), I think an exact RubyGems version match is always necessary at the moment?

File: lib/rubygems/commands/rebuild_command.rb
Comment (deivid-rodriguez) [inline lib/rubygems/commands/rebuild_command.rb:82]: Do we want to include the full output of the current implementation in the command documentation/description? I think it's likely to get out of date and become confusing?

File: lib/rubygems/commands/rebuild_command.rb
Comment (deivid-rodriguez) [inline lib/rubygems/commands/rebuild_command.rb:140]: Should default to building things in `/tmp`? We print paths for inspection and that way we don't leave artifacts around in the working directory. Or maybe current directory's `tmp/` if it exists?

File: lib/rubygems/commands/rebuild_command.rb
Comment (deivid-rodriguez) [inline lib/rubygems/commands/rebuild_command.rb:240]: What's different with what `gem fetch` does? Could we extract some common method for both, or directly reuse `FetchCommand` here?

**Review by deivid-rodriguez — COMMENTED** [review]
I like this!

I worry that having to use the exact same RubyGems version makes it harder to reproduce builds? If only that metadata field is different, things are safe, right? We could consider build with `rubygems_version` set to the one that originally built the gem, and if they match then, inform about that?

File: lib/rubygems/commands/rebuild_command.rb
Comment (segiddins) [reply lib/rubygems/commands/rebuild_command.rb:55]: yes -- one thing we could do is error out ahead of time if the wrong gem version is used, or add a way to override the rubygems version serialized into the gemspec, the same way we allow overriding the source date epoch?

Comment (deivid-rodriguez) [reply lib/rubygems/commands/rebuild_command.rb:55]: Yeah. I'd go with the latter, seems like it could make this more useful. Also I see some other differences in metadata, like some trailing spaces differences caused by the rubygems version divergence. We could also normalize those.

**Comment by duckinator** [comment]
> I like this!

I'm glad. I've put a lot of effort into it. 😂 

> I worry that having to use the exact same RubyGems version makes it harder to reproduce builds?

It definitely does.

> If only that metadata field is different, things are safe, right?

Yes.

> We could consider build with rubygems_version set to the one that originally built the gem, and if they match then, inform about that?

Yeah, this is definitely an option worth considering. And if this is included in the metadata (as @segiddins suggested), we could check RubyGems version before trying the build and just print a message like "It is recommended to use RubyGems v$VERSION for rebuilding this gem" or something like that.

File: lib/rubygems/commands/rebuild_command.rb
Comment (duckinator) [reply lib/rubygems/commands/rebuild_command.rb:240]: The main reason I can't use `FetchCommand#execute` is that I need a clear indicator of whether the lookup + download failed.

I'm basically using a modified version of this part, which downloads a single gem:

https://github.com/rubygems/rubygems/blob/bb66253/lib/rubygems/commands/fetch_command.rb#L73-L91

File: lib/rubygems/commands/rebuild_command.rb
Comment (duckinator) [reply lib/rubygems/commands/rebuild_command.rb:150]: This is definitely a good idea.

File: lib/rubygems/commands/rebuild_command.rb
Comment (duckinator) [reply lib/rubygems/commands/rebuild_command.rb:222]: Yeah, it looks like it's identical, so it can be shared.

Any idea where that should be put? It seems like shared things in `rubygems/commands/` just get dumped in `rubygems/*`, which seems like a pattern that may be worth worth breaking. Perhaps creating `rubygems/commands/shared/` for that purpose? (I'm okay with other names for such a directory, it's just first one that came to mind.)

<details><summary>details about shared things for `rubygems/commands/*`</summary>

For context, it appears that at least these files are only used in `rubygems/commands/` (and corresponding tests):

- `rubygems/gemcutter_utilities`
- `rubygems/install_message`
- `rubygems/install_default_message`
- `rubygems/install_update_options`
- `rubygems/installer`
- `rubygems/local_remote_options`
- `rubygems/query_utils`
- `rubygems/rdoc`

</details>

File: lib/rubygems/commands/rebuild_command.rb
Comment (duckinator) [reply lib/rubygems/commands/rebuild_command.rb:156]: Done. :3

File: lib/rubygems/commands/rebuild_command.rb
Comment (duckinator) [reply lib/rubygems/commands/rebuild_command.rb:86]: It probably makes sense to pull this out and include it (with some other information) on <https://guides.rubygems.org>.

File: lib/rubygems/commands/rebuild_command.rb
Comment (duckinator) [reply lib/rubygems/commands/rebuild_command.rb:147]: It now uses a temporary directory, instead of the working directory.

File: lib/rubygems/commands/rebuild_command.rb
Comment (duckinator) [reply lib/rubygems/commands/rebuild_command.rb]: Resolved by making `gem rebuild` use a temporary directory for output.

File: lib/rubygems/commands/rebuild_command.rb
Comment (duckinator) [reply lib/rubygems/commands/rebuild_command.rb:87]: This was mainly there so I had a practical example to reference. I think <https://smallest.dog/resources/reproducible-gem-builds/> has taken over that purpose, however, so I'm removing it from the help text.

File: lib/rubygems/commands/rebuild_command.rb
Comment (duckinator) [reply lib/rubygems/commands/rebuild_command.rb]: I updated the help text to be more correct.

**Comment by duckinator** [comment]
@segiddins @deivid-rodriguez I've done a lot of cleanup on this today. From your reviews, this is all that's left:

1. `find_gemspec()` is redundant with the `gem build` command.
2. `download_gem()` and `FetchCommand#execute` could potentially be refactored to have a common helper function.
3. It does not bail early if the RubyGems version is wrong.

Do y'all think any of these should block merging this PR? And do you have any other feedback?

As for resolving those 3 remaining things:
- For 1 and 2, I'm not sure where to put the common functionality.
- For 3, I'm not sure how to take a path to a `.gem` file and get metadata from it.

**Comment by duckinator** [comment]
Rebased off master (2e28234).

File: lib/rubygems/commands/rebuild_command.rb
Comment (segiddins) [inline lib/rubygems/commands/rebuild_command.rb:105]:
```suggestion
    File.open(file, Gem.binary_mode) do |f|
```

File: lib/rubygems/commands/rebuild_command.rb
Comment (segiddins) [inline lib/rubygems/commands/rebuild_command.rb:100]:
```suggestion
    Digest::SHA256.hexdigest(Gem.read_binary(file))
```

though ideally we'd use one of the helpers @martinemde wrote, so we can avoid reading the whole gem contents into memory at once

File: lib/rubygems/commands/rebuild_command.rb
Comment (segiddins) [reply lib/rubygems/commands/rebuild_command.rb:228]: could it be extracted into a common `SpecFetching` module?

**Review by segiddins — COMMENTED** [review]
> For 1 and 2, I'm not sure where to put the common functionality.

Just throw it into a new module somewhere

> For 3, I'm not sure how to take a path to a .gem file and get metadata from it.

`Gem::Package.new(path).spec` will give you the gemspec object from the .gem at path

**Comment by duckinator** [comment]
It now bails early if the RubyGems version doesn't match:

```
puppy@orthrus:~/test/okay$ gem rebuild okay 12.0.2
Fetching okay-12.0.2.gem
Downloaded okay version 12.0.2 as /tmp/gem_rebuild20240131-79494-x86r8z/old/okay-12.0.2.gem.
ERROR:  You need to use the same RubyGems version okay v12.0.2 was built with.

okay v12.0.2 was built using RubyGems v3.5.0.dev.
Gem files include the version of RubyGems used to build them.
This means in order to reproduce okay-12.0.2.gem, you must also use RubyGems v3.5.0.dev.

Please install RubyGems v3.5.0.dev and try again.
puppy@orthrus:~/test/okay$
```

**Comment by duckinator** [comment]
Rebased off master (5602c2b).

**Comment by duckinator** [comment]
@segiddins @deivid-rodriguez I believe all that's left from your suggestions is:

- [ ] Extract `find_gemspec()` into a common module and have `RebuildCommand` and `BuildCommand` both use it.
- [ ] Extract `download_gem()` into a common module and have `RebuildCommand` and `BuildCommand` both use it.

I'm not feeling great atm, so I'm gonna call it a day. If you want to leave these two things for a follow-up PR, you can go ahead and merge this; otherwise, I'll try to get those changes done sometime in the next week or so.

File: lib/rubygems/commands/rebuild_command.rb
Comment (deivid-rodriguez) [inline lib/rubygems/commands/rebuild_command.rb:157]: Should we add an `else` here

> To see package differences, use the `--diff` flag (needs `diffoscope` installed)

Comment (duckinator) [reply lib/rubygems/commands/rebuild_command.rb:157]: I have mixed feelings about that. On one hand, it's good for discoverability; on the other hand, having the clear SUCCESS/FAILURE right at the end by default has proven really nice in practice.

Comment (duckinator) [reply lib/rubygems/commands/rebuild_command.rb:159]: Added a message suggesting using `--diff` if you didn't use it.

**Comment by duckinator** [comment]
Extracting common code from `RebuildCommand#download_gem` and `FetchCommand#execute` doesn't seem to make sense. There's only like 4 lines that are the same, and they're scattered throughout the two functions.

I believe everything else y'all have requested has been addressed.

**Comment by duckinator** [comment]
Rebased off master (1aa7252). @segiddins @deivid-rodriguez this is ready for a final review. If y'all are happy with it and CI passes, it's ready to merge.

**Review by segiddins — APPROVED** [review]

**Review by deivid-rodriguez — APPROVED** [review]

**Comment by duckinator** [comment]
Rebased off master (5956bfb). Will merge if CI passes. :+1:

File: lib/rubygems/commands/rebuild_command.rb
Comment (nobu) [inline lib/rubygems/commands/rebuild_command.rb:13]: What is this constant for?
Seems unused.

Comment (duckinator) [reply lib/rubygems/commands/rebuild_command.rb:13]: @nobu good catch, it is unused. I opened #7546 to remove it. :+1:

**Comment by duckinator** [comment]
@deivid-rodriguez I think we should release it. The plugin is basically a way to "backport" the functionality to old versions, and only gets updated by [a script that updates the plugin's files with ones from the RubyGems repo](https://github.com/duckinator/gem-rebuild/blob/main/bin/update-source-files).

The sooner the real version is out, the sooner the plugin becomes irrelevant.

**Comment by nobu** [comment]
Is it intentional that the downloaded and rebuilt gem files are left under `gem_rebuild*` temporary directory?

**Comment by deivid-rodriguez** [comment]
mmmm, I think it'd be better to clean after ourselves? Thoughts @duckinator?

**Comment by duckinator** [comment]
leaving the files after `gem rebuild` was intentional, for local inspection, but the _test suite_ should be made to clean up after itself independently of that.

**Pull request merged**
